cacheops/transaction.py

```diff
@@ -0,0 +1,185 @@
+        return_value = _find_second_latest_context(latest)
+        if return_value is _marker:
+            return latest
+        if return_value is not None:
+            return return_value
+        return _marker
+    return None
+
+
+def _drop_latest_context(contexts):
+    if not contexts:
+        return True
+    latest = contexts[-1]
+    if isinstance(latest, list):
+        return_value = _drop_latest_context(latest)
+        if return_value:
+            contexts.pop()
+            return False
+        return return_value
+    return True
+
+
+def _flatten_and_ignore_none(items):
+    """
+    returns all non list items in a list of lists of of arbitrary depth at a single depth,
+     ignoring non list items that are `None`.
+    """
+    for item in items:
+        if isinstance(item, list):
+            # todo: use yield from in python 3
+            for x in _flatten_and_ignore_none(item):
+                yield x
+        elif item is not None:
+            yield item
+
+
+class TransactionQueue(threading.local):
+    """
+    `_queue` is a list of lists of of arbitrary depth. it represents a queue of items, grouped by
+     nested contexts. non-list elements at each level are items in the queue. list elements are
+     considered contexts. the latest context is the deepest list who's right most boundary is also
+     the right most boundary of the outer most context. `None` items are added to the list, in order
+     to make second latest contexts the new latest context (when an inner most context ends). These
+     `None` objects are not return on __iter__.
+
+     You probably want to sub class this class to override commit_transaction, or in alternate cases
+      rollback_transaction.
+    """
+    def __init__(self):
+        self._queue = None
+
+    def start_transaction(self):
+        self._queue = []
+
+    def commit_transaction(self):
+        self._queue = None
+
+    def rollback_transaction(self):
+        self._queue = None
+
+    def start_savepoint(self):
+        latest_context = _find_latest_context(self._queue)
+        if latest_context is None:
+            latest_context = self._queue
+        latest_context.append([])
+
+    def commit_savepoint(self):
+        second_latest = _find_second_latest_context(self._queue)
+        if second_latest is None or second_latest is _marker:
+            second_latest = self._queue
+        second_latest.append(None)
+
+    def rollback_savepoint(self):
+        return_value = _drop_latest_context(self._queue)
+        if return_value:
+            raise ValueError('Trying to drop the top most context.')
+
+    def in_transaction(self):
+        return self._queue is not None
+
+    def append(self, item):
+        latest_context = _find_latest_context(self._queue)
+        if latest_context is None:
+            latest_context = self._queue
+        latest_context.append(item)
+
+    def __iter__(self):
+        return _flatten_and_ignore_none(self._queue)
```

Comment:
Simplify this class interface to `.begin()`, `.commit()` and `.rollback()`. Nested begins are savepoints. Each level has a list in `.queue`, so:
```python
assert queue._queue = []

queue.begin()
assert queue._queue = [[]]

queue.append(call1)
queue.append(call2)
queue.begin()
queue.append(call3)
queue.begin()

assert queue._queue = [[call1, call2], [call3], []]
```
On rollback we just drop last item, on commit we concat last item to second last and then drop last one:
```python
self._queue[-2] += self.pop()
```
When we commit highest level we call `.apply()`:
```python
for func, args, kwargs in self._queue[0]:
    func(*args, **kwargs)
```